### PR TITLE
New Jenkinsfile Setup

### PR DESCRIPTION
**Pull request by tskinn**
This is part of https://github.com/department-of-veterans-affairs/vets.gov-team/issues/16424

This PR moves logic for doing builds/prearchives/archives to a new common file to be shared between the normal Jenkinsfile and Jenkinsfile.content which is for content only builds.

This must be merged with https://github.com/department-of-veterans-affairs/devops/pull/4241 which creates the content-only build job in jenkins and updates the daily release to use the content-only build.

### CODE REVIEW

File: src/site/stages/build/plugins/download-assets.js [hunk 1]
```diff
@@ -74,7 +74,11 @@ async function downloadFromLiveBucket(files, buildOptions) {
 async function downloadFromArchive(files, assetSource, buildtype) {
   const archiveUrl = `https://s3-us-gov-west-1.amazonaws.com/vetsgov-website-builds-s3-upload/${assetSource}/${buildtype}.tar.bz2`;
 
-  const localPath = path.join(__dirname, '../../../../../temp/site.tar.bz2');
+  const localPath = path.join(
+    __dirname,
+    `../../../../../temp/${buildtype}/site.tar.bz2`,
+  );
+
```
Comment (tskinn) [inline src/site/stages/build/plugins/download-assets.js]: Since we are trying to do the builds in parallel, I think this was causing some issues. This addition just introduces a difference between the builds.

File: Jenkinsfile [hunk 2]
```diff
@@ -2,159 +2,56 @@ import org.kohsuke.github.GitHub
...
+    checkout scm
+    ref = sh(returnStdout: true, script: 'git rev-parse HEAD').trim()
+  }
 
-        sh "mkdir -p build"
-        sh "mkdir -p logs/selenium"
-        sh "mkdir -p coverage"
-        sh "mkdir -p temp"
+  def buildUtil = load "vets-website/Jenkinsfile.common"
+  def dockerArgs = "-v ${WORKSPACE}/vets-website:/application -v ${WORKSPACE}/vagov-content:/vagov-content"
```
[the remark this answers is not in the document]
Comment (wyattwalter) [reply Jenkinsfile]: Since these have to be passed to various stages, it may make sense to move this logic to set the string in the common file, that way the args definition isn't duplicated.

File: Jenkinsfile [hunk 3]
```diff
@@ -2,159 +2,56 @@ import org.kohsuke.github.GitHub
...
+  def cmsEnv = params.get('cmsEnv', 'none')
+  def imageTag = java.net.URLDecoder.decode(env.BUILD_TAG).replaceAll("[^A-Za-z0-9\\-\\_]", "-")
```
Comment (wyattwalter) [inline Jenkinsfile]: imageTag could also move into the common file

File: Jenkinsfile (hunk 2 above)
[the remark this answers is not in the document]
Comment (tskinn) [reply Jenkinsfile]: Yeah the pwd() was during the wild goose chase trying to fix docker. Good call.

[the remark this answers is not in the document]
Comment (tskinn) [reply Jenkinsfile]: @wyattwalter should I just get rid of most if not all of the parameters for these functions that run stages? Since most of the needed info is in the common.groovy file anyways.

[the remark this answers is not in the document]
Comment (wyattwalter) [reply Jenkinsfile]: I think that would be ideal. I know that the main Jenkinsfile does need at least one of those parameters, though :(

File: jenkins/common.groovy [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+}
+
+def isDeployable() {
+  return (IS_DEV_BRANCH ||
+          IS_STAGING_BRANCH) &&
+    !env.CHANGE_TARGET &&
+    !currentBuild.nextBuild // if there's a later build on this job (branch), don't deploy
+}
+
+def shouldBail() {
```
Comment (jbalboni) [inline jenkins/common.groovy]: It seems like there might be a more descriptive name for this function, since we bail out of certain stages for multiple reasons. `hasLaterBuild` or something?

**Review by jbalboni — APPROVED** [review]
Seems like an improvement to me, though it's a little hard to follow all the changes.

File: jenkins/common.groovy (hunk 4 above)
Comment (tskinn) [reply jenkins/common.groovy]: Yeah and it seems like `isReviewable()` has similar logic that could be reused/renamed. I'll do another pass and try to make things more decipherable.

Comment (tskinn) [reply jenkins/common.groovy]: I've been looking over the logic in these few functions that contain logic about whether to bail or not and I'm not sure if they make sense to me anymore or I've just been looking at this for too long. Why is `!IS_DEV_BRANCH && !IS_STAGING_BRANCH && !IS_PROD_BRANCH` included in the shouldbail function?

Comment (ncksllvn) [reply jenkins/common.groovy]: `shouldBail` is a weird one, and should probably be separated out into separate conditions or at least a more expressive function. 

My understanding is that if we're on a dev/staging/prod branch (which in our case is always master but sometimes there are special circumstances where we build dev off of a feature branch instead), we always want our build to complete and be archived into S3. 

However, if we're on a different branch (anything other than master) then we're likely in a feature branch/pull request and pushing new updates regularly. In this case, if a new commit is pushed before the former is finished building, we cancel the old job. I believe we do this because we aren't concerned about archiving those intermediary commits on feature branches, and it conserves resources on Jenkins. This condition is detected by ` currentBuild.nextBuild` - if that's true, then there's a more recent commit on the Jenkins queue.

Comment (wyattwalter) [reply jenkins/common.groovy]: That was to protect us from aborting the build on the master branch. There's an option to set a branch for each environment (defaults to master for all), so we check each of them.

File: jenkins/common.groovy [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+  }
+}
+
+def setup(String dockerTag, String dockerArgs) {
+  stage("Setup") {
+
+    buildTypeOverride = DRUPAL_MAPPING.get(params.get('cmsEnv', 'none'), null)
+    if(buildTypeOverride) {
+      VAGOV_BUILDTYPES = [buildTypeOverride]
+    }
```
Comment (ncksllvn) [inline jenkins/common.groovy]: Does 'buildTypeOverride' have to be applied here, or can it be applied at the top-level scope? Just thinking it might be clearer to initialize `VAGOV_BUILDTYPES` like -

```groovy
ALL_VAGOV_BUILDTYPES = 
  'vagovdev',
  'vagovstaging',
  'vagovprod'
]

BUILD_TYPE_OVERRIDE = DRUPAL_MAPPING.get(params.get('cmsEnv', 'none'), null)

VAGOV_BUILDTYPES = BUILD_TYPE_OVERRIDE ? [BUILD_TYPE_OVERRIDE]  : ALL_VAGOV_BUILDTYPES
```

File: Jenkinsfile [hunk 6]
```diff
@@ -163,58 +63,13 @@ node('vetsgov-general-purpose') {
...
-      throw error
-    }
-  }
+  withCms = (cmsEnv != 'none' && cmsEnv != 'live')
+  buildUtil.build(ref, dockerImage, dockerArgs, withCms)
 
   // Run E2E and accessibility tests
   stage('Integration') {
-    if (shouldBail() || !VAGOV_BUILDTYPES.contains('vagovprod')) { return }
+    if (buildUtil.shouldBail() || !VAGOV_BUILDTYPES.contains('vagovprod')) { return }
```
Comment (ncksllvn) [inline Jenkinsfile]: I think we could make these `VAGOV_BUILDTYPES.contains(` conditions clearer, because I had to trace back where `VAGOV_BUILDTYPES` is modified to understand what was happening there. It might be clearer to create a new lowercase variable `targetBuildtypes` to help show that this value is no longer constant. I left another comment related to this.

File: Jenkinsfile [hunk 7]
```diff
@@ -298,32 +105,26 @@ node('vetsgov-general-purpose') {
...
-        script {
-          commit = sh(returnStdout: true, script: "git rev-parse HEAD").trim()
-        }
-      }
+      if (!stages.isDeployable()) { return }
 
-      if (IS_DEV_BRANCH && VAGOV_BUILDTYPES.contains('vagovdev')) {
-        runDeploy('deploys/vets-website-dev', commit)
-        runDeploy('deploys/vets-website-vagovdev', commit)
+      if (stagess.IS_DEV_BRANCH && stagess.VAGOV_BUILDTYPES.contains('vagovdev')) {
```
Comment (ncksllvn) [inline Jenkinsfile]: Should there be two S's here, `stagess`?

**Review by ncksllvn — APPROVED** [review]
Great work, really nice improvement!

File: Jenkinsfile.common [hunk 8]
```diff
@@ -0,0 +1,180 @@
...
+    slackSend message: message,
+      color: 'danger',
+      failOnError: true
+  }
+}
+
+def setup(String dockerTag, String dockerArgs) {
+  stage("Setup") {
+
+    buildTypeOverride = DRUPAL_MAPPING.get(params.get('cmsEnv', 'none'), null)
```
[the remark this answers is not in the document]
Comment (tskinn) [reply Jenkinsfile.common]: I think we are going to keep the cmsEnv param for now so that drupal can publish dev and staging on demand.
